File: step_03_arrays/3.1_easy/P007_move_zeros_to_end/Solution.java
/**
 * Problem: Move Zeros to End (LeetCode #283)
 * Difficulty: EASY | XP: 10
 *
 * Move all zeros to the end of the array while maintaining
 * the relative order of non-zero elements. In-place.
 *
 * @author DSA_Nova
 */
import java.util.Arrays;

// ============================================================
// Approach 1: Brute Force (Extra Array)
// Time: O(n) | Space: O(n)
// ============================================================
class BruteForce {
    public static void moveZeroes(int[] nums) {
        int n = nums.length;
        int[] temp = new int[n]; // defaults to all zeros
        int idx = 0;

        // Copy non-zeros
        for (int num : nums) {
            if (num != 0) {
                temp[idx++] = num;
            }
        }
        // remaining positions are already 0

        System.arraycopy(temp, 0, nums, 0, n);
    }
}

// ============================================================
// Approach 2: Optimal (Two-Pointer Write + Backfill)
// Time: O(n) | Space: O(1)
// ============================================================
class Optimal {
    public static void moveZeroes(int[] nums) {
        int writeIdx = 0;

        // Place all non-zeros at the front
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                nums[writeIdx++] = nums[i];
            }
        }

        // Fill the rest with zeros
        while (writeIdx < nums.length) {
            nums[writeIdx++] = 0;
        }
    }
}

// ============================================================
// Approach 3: Best (Swap Variant -- single pass, no backfill)
// Time: O(n) | Space: O(1)
// ============================================================
class Best {
    public static void moveZeroes(int[] nums) {
        int writeIdx = 0;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                // Swap nums[i] with nums[writeIdx]
                int temp = nums[i];
                nums[i] = nums[writeIdx];
                nums[writeIdx] = temp;
                writeIdx++;
            }
        }
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Move Zeros to End ===\n");

        int[][] testCases = {
            {0, 1, 0, 3, 12},
            {0},
            {1, 2, 3},
            {0, 0, 0, 1},
            {0, 0, 0},
            {-1, 0, 3, 0, -5}
        };
        int[][] expected = {
            {1, 3, 12, 0, 0},
            {0},
            {1, 2, 3},
            {1, 0, 0, 0},
            {0, 0, 0},
            {-1, 3, -5, 0, 0}
        };

        for (int t = 0; t < testCases.length; t++) {
            int[] a1 = testCases[t].clone();
            int[] a2 = testCases[t].clone();
            int[] a3 = testCases[t].clone();

            BruteForce.moveZeroes(a1);
            Optimal.moveZeroes(a2);
            Best.moveZeroes(a3);

            boolean pass = Arrays.equals(a1, expected[t])
                        && Arrays.equals(a2, expected[t])
                        && Arrays.equals(a3, expected[t]);

            System.out.println("Input:    " + Arrays.toString(testCases[t]));
            System.out.println("Brute:    " + Arrays.toString(a1));
            System.out.println("Optimal:  " + Arrays.toString(a2));
            System.out.println("Best:     " + Arrays.toString(a3));
            System.out.println("Expected: " + Arrays.toString(expected[t]));
            System.out.println("Pass:     " + pass);
            System.out.println();
        }
    }
}

File: step_03_arrays/3.1_easy/P007_move_zeros_to_end/solution.py
"""
Problem: Move Zeros to End (LeetCode #283)
Difficulty: EASY | XP: 10

Move all zeros to the end of the array while maintaining
the relative order of non-zero elements. In-place.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE (Extra Array)
# Time: O(n) | Space: O(n)
# ============================================================
def brute_force(nums: List[int]) -> List[int]:
    """Create a new array: non-zeros first, then zeros."""
    result = [x for x in nums if x != 0]
    result.extend([0] * (len(nums) - len(result)))
    return result


# ============================================================
# APPROACH 2: OPTIMAL (Two-Pointer Write + Backfill)
# Time: O(n) | Space: O(1)
# ============================================================
def optimal(nums: List[int]) -> List[int]:
    """Write pointer places non-zeros, then backfill zeros."""
    result = nums[:]
    write_idx = 0

    for i in range(len(result)):
        if result[i] != 0:
            result[write_idx] = result[i]
            write_idx += 1

    # Fill remaining with zeros
    for i in range(write_idx, len(result)):
        result[i] = 0

    return result


# ============================================================
# APPROACH 3: BEST (Swap Variant -- single pass)
# Time: O(n) | Space: O(1)
# ============================================================
def best(nums: List[int]) -> List[int]:
    """Swap non-zeros to the front in one pass."""
    result = nums[:]
    write_idx = 0

    for i in range(len(result)):
        if result[i] != 0:
            result[i], result[write_idx] = result[write_idx], result[i]
            write_idx += 1

    return result


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Move Zeros to End ===\n")

    test_cases = [
        ([0, 1, 0, 3, 12], [1, 3, 12, 0, 0]),
        ([0], [0]),
        ([1, 2, 3], [1, 2, 3]),
        ([0, 0, 0, 1], [1, 0, 0, 0]),
        ([0, 0, 0], [0, 0, 0]),
        ([-1, 0, 3, 0, -5], [-1, 3, -5, 0, 0]),
    ]

    for nums, expected in test_cases:
        b = brute_force(nums)
        o = optimal(nums)
        be = best(nums)
        status = "PASS" if b == expected and o == expected and be == expected else "FAIL"
        print(f"Input:    {nums}")
        print(f"  Brute:    {b}")
        print(f"  Optimal:  {o}")
        print(f"  Best:     {be}")
        print(f"  Expected: {expected}  [{status}]\n")
